### PR TITLE
Use non-blocking write in message_job()

### PR DESCRIPTION
**Pull request by prakashcv13**
#### Describe Bug or Feature
* [PP-1000](https://pbspro.atlassian.net/browse/PP-1000)

When an interactive job sends a lot of spam to stdout/stderr, MoM hangs trying to kill the job (on exceeding walltime).

#### Describe Your Change
When MoM decides to kill the job - it calls message_job() that does a BLOCKING write() to display a message telling why the job is being killed. The file that the MoM writes to is a pseudo-tty (for an interactive job) that is being used by the child pbs_mom (job process) to send data to qsub.
As there is a lot of data being sent - there is a chance that a lot of data is buffered and until all the data is being read by the other side, the MoM process that called message_job is blocked.

As a solution - we use O_NONBLOCK in message_job().

To reproduce - I am using the cgroup freezer subsystem to freeze qsub processes so that the buffering happens. It is not straightforward to do that through PTL - so I am attaching manual test logs.

Test Logs
[after_fix.txt](https://github.com/PBSPro/pbspro/files/3336592/after_fix.txt)
[before_fix.txt](https://github.com/PBSPro/pbspro/files/3336593/before_fix.txt)
[mom_valgrind.txt](https://github.com/PBSPro/pbspro/files/3349310/mom_valgrind.txt)

### CODE REVIEW

**Comment by prakashcv13** [comment]
Hi All, This PR is open for review.

File: src/resmom/requests.c [hunk 1]
```diff
@@ -765,15 +783,37 @@ message_job(job *pjob, enum job_file jft, char *text)
 		pstr[len++] = '\n';	/* append new-line */
 		text = pstr;
 	}
-	(void)write(fds, text, len);
 #ifdef	WIN32
+	(void)write(fds, text, len);
 	(void)_commit(fds);
+#else
+	flags = fcntl(fds, F_GETFL, 0);
+	fcntl(fds, F_SETFL, flags | O_NONBLOCK);
```
Comment (riyazhakki) [inline src/resmom/requests.c]: fds is already opened with O_NONBLOCK flag, why to set it again?

File: src/resmom/requests.c [hunk 2]
```diff
@@ -765,15 +783,37 @@ message_job(job *pjob, enum job_file jft, char *text)
...
+			}
+		}
+	}
 #endif
 	(void)close(fds);
 	if (pstr)
 		free(pstr);
 
-	return PBSE_NONE;
+	if (total_bytes_written == len)
```
Comment (riyazhakki) [inline src/resmom/requests.c]: What if above write() call returns -1 (in case errno=EAGAIN) at least one time in the loop ? Does this condition satisfy on next successful write?

File: src/resmom/requests.c [hunk 3]
```diff
@@ -746,16 +750,30 @@ message_job(job *pjob, enum job_file jft, char *text)
...
 	/* set to append mode */
 	SetFilePointer((HANDLE)_get_osfhandle(fds), (LONG)NULL,
 		(PLONG)NULL, FILE_END);
-#endif
+#else
+	while ((slept < 5) && (fds < 0)) {
+		fds = open_std_file(pjob, jft, O_WRONLY | O_APPEND | O_NONBLOCK,
+	                         pjob->ji_qs.ji_un.ji_momt.ji_exgid);
+		if (fds < 0) {
+			if ( errno == EAGAIN || errno == EWOULDBLOCK) {
```
[the remark this answers is not in the document]
Comment (prakashcv13) [reply src/resmom/requests.c]: seteuid() and seteguid() do not set errno to EAGAIN. However, setresuid() and setresgid() do. If any of these calls failed and set errno to EAGAIN, the attempt to open the file was not successful as open() is not called at all. We would need to retry as EAGAIN is a temporary error.
It is my view, it really does not matter which system call failed and set the errno to EAGAIN in this case.

File: src/resmom/requests.c [hunk 4]
```diff
@@ -746,16 +750,30 @@ message_job(job *pjob, enum job_file jft, char *text)
...
+				sleep(1);
```
[the remark this answers is not in the document]
Comment (prakashcv13) [reply src/resmom/requests.c]: yes, the mom will be blocked if the errno is set to EAGAIN/EWOULDBLOCK, which is better than being blocked forever. We can use usleep() to sleep for 0.5 seconds, to reduce the amount of time mom will be blocked.

File: src/resmom/requests.c (hunk 2 above)
Comment (prakashcv13) [reply src/resmom/requests.c]: yes.

File: src/resmom/requests.c [hunk 5]
```diff
@@ -765,15 +783,37 @@ message_job(job *pjob, enum job_file jft, char *text)
...
+	(void)write(fds, text, len);
 	(void)_commit(fds);
+#else
+	flags = fcntl(fds, F_GETFL, 0);
+	fcntl(fds, F_SETFL, flags | O_NONBLOCK);
+
+	while ((slept < 5) && (total_bytes_written < len)) {
+		bytes_written = write(fds, text, len-total_bytes_written);
+		total_bytes_written += bytes_written;
+		text += bytes_written;
```
[the remark this answers is not in the document]
Comment (prakashcv13) [reply src/resmom/requests.c]: how?

[the remark this answers is not in the document]
Comment (riyazhakki) [reply src/resmom/requests.c]: On write failure bytes_written will have -1 and the line 796 'text += bytes_written' will move text pointer to one location backward.

**Comment by prakashcv13** [comment]
@riyazhakki - I have addressed your review comment, please take a look.

File: src/resmom/requests.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (mike0042) [reply src/resmom/requests.c]: I tried using WORK_Immed to instantiate the PMIx server process and noticed a delay of around a second before the work task actually started. This was unacceptable for a rapid-fire launch scenario. The call to open_std_file() should return success the first time in most cases. I don't think we need to try to open the file more than three times, pausing 0.25 seconds between attempts for a total of 0.5 seconds of sleep. Also, this should be a for loop to instantiate and increment slept.

File: src/resmom/requests.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (arungrover) [reply src/resmom/requests.c]: I disagree, we are trying to find a problem in one call and checking errno which can potentially be set by another system call.
In function open_file_as_user() open is called before calling setresuid/setresgid. This means if open fails for some reason you do not want to retry, this code might end up retrying even in that case because it will be working off of errno set by setresuid/setresgid calls.

**Comment by prakashcv13** [comment]
@mike0042 and @arungrover - I have tried to address your comments, requesting another look.

File: src/resmom/requests.c [hunk 6]
```diff
@@ -757,16 +756,18 @@ message_job(job *pjob, enum job_file jft, char *text)
...
+	unsigned int usecs = 250 * 1000; /* 250 milliseconds */
+	for (i = 0; i < 3; i++) {
 		fds = open_std_file(pjob, jft, O_WRONLY | O_APPEND | O_NONBLOCK,
 	                         pjob->ji_qs.ji_un.ji_momt.ji_exgid);
-		if (fds < 0) {
-			if (errno == EAGAIN || errno == EWOULDBLOCK) {
-				sleep(1);
-				slept++;
-			} else
+		if (fds > 0)
```
Comment (arungrover) [inline src/resmom/requests.c]: Was this change intentional? If fd is > 0 then that would mean open() succeeded but we will still retry if errno is set to EAGAIN or EWOULDBLOCK.

Comment (prakashcv13) [reply src/resmom/requests.c]: not intentional. I have pushed the correct changes now.

File: src/resmom/start_exec.c [hunk 7]
```diff
@@ -6379,18 +6379,23 @@ starter_return(int upfds, int downfds, int code, struct startjob_rtn *sjrtn)
...
+	int open_errno = 0; 
 	extern gid_t pbsgroup;
 
 	/* must open or create file as the user */
 
 	if (impersonate_user(exuid, exgid) == -1)
 		return -1;
 
-	fds = open(path, oflag, mode);
+	if ((fds = open(path, oflag, mode)))
```
Comment (arungrover) [inline src/resmom/start_exec.c]: Please set errno only when open returns -1

File: src/resmom/requests.c [hunk 8]
```diff
@@ -793,6 +793,8 @@ message_job(job *pjob, enum job_file jft, char *text)
 				usleep(usecs);
 			else {
 				(void)close(fds);
+				if (pstr)
+					free(pstr);
```
Comment (mike0042) [inline src/resmom/requests.c]: You have properly initialized pstr to NULL when you declared it, so no need to check for NULL. Just call free().

File: src/resmom/start_exec.c (hunk 7 above)
Comment (prakashcv13) [reply src/resmom/start_exec.c]: done

File: src/resmom/requests.c (hunk 8 above)
Comment (prakashcv13) [reply src/resmom/requests.c]: done.

**Comment by prakashcv13** [comment]
Hi @arungrover and @mike0042 - Thank you for the comments, I have addressed the latest review comments, requesting another look.

File: src/resmom/requests.c [hunk 9]
```diff
@@ -746,34 +747,71 @@ message_job(job *pjob, enum job_file jft, char *text)
...
 		(void)strcpy(pstr, text);
 		pstr[len++] = '\n';	/* append new-line */
 		text = pstr;
 	}
-	(void)write(fds, text, len);
 #ifdef	WIN32
+	(void)write(fds, text, len);
 	(void)_commit(fds);
+#else
+	for (i = 0; i < 5; i++) {
```
Comment (mike0042) [inline src/resmom/requests.c]: Final comment... I think this should be three tries as well.

Comment (prakashcv13) [reply src/resmom/requests.c]: ok - let us have only three tries...

**Comment by prakashcv13** [comment]
thank you @arungrover and @mike0042 for the review and approvals, shall we take this PR forward.
